syntax = "proto3";

package quo.rt;

// Descriptor for a module.
message ModuleDescriptor {
  // Members.
  repeated ModuleMemberDescriptor member_descs = 1;
}

// Descriptor for a module member.
message ModuleMemberDescriptor {
  oneof type {
    ClassDescriptor class_desc = 1;
    FnDescriptor fn_desc = 2;
  }
}

// Descriptor for a class.
message ClassDescriptor {
  // Qualified name of the class.
  string name = 1;
  // Members.
  repeated ClassMemberDescriptor member_descs = 2;
}

// Descriptor for a class member.
message ClassMemberDescriptor {
  // Qualified name of the class that declared this member.
  string decl_class_name = 1;
  // Fields.
  oneof type {
    PropDescriptor prop_desc = 2;
    FnDescriptor fn_desc = 3;
  }
}

// Descriptor for a property (instance variable).
message PropDescriptor {
  // Qualified name of the function.
  string name = 1;
  // Index of this property within the object's memory layout.
  int32 index = 2;
  // Qualified name of the type of this property.
  string type_name = 3;
  // Reference mode of this property.
  RefMode ref_mode = 4;
};

// Descriptor for a method.
message FnDescriptor {
  // Qualified name of the function.
  string name = 1;
  // Index of this method in the function table.
  int32 fn_table_index = 2;
  // Parameters.
  repeated FnParamDescriptor param_descs = 3;
  // Qualified name of the return type of the function.
  string return_type_name = 4;
  // Whether this was an override function.
  bool is_override = 5;
}

message FnParamDescriptor {
  // Name of the parameter.
  string name = 1;
  // Reference mode.
  RefMode ref_mode = 2;
  // Qualified name of the type of the paramter.
  string type_name = 3;
}

// Mode of a reference.
enum RefMode {
  STRONG_REF = 0;
  WEAK_REF = 1;
}
